app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,233 @@
+from __future__ import annotations
+
+import textwrap
+from datetime import datetime
+from typing import cast
+from typing import TypedDict
+
+import app.state.services
+from app._typing import _UnsetSentinel
+from app._typing import UNSET
+
+# +--------------+------------------------+------+-----+---------+-------+
+# | Field        | Type                   | Null | Key | Default | Extra |
+# +--------------+------------------------+------+-----+---------+-------+
+# | id           | int                    | NO   | PRI | NULL    |       |
+# | userid       | int                    | NO   |     | NULL    |       |
+# | ip           | varchar(45)            | NO   |     | NULL    |       |
+# | osu_ver      | date                   | NO   |     | NULL    |       |
+# | osu_stream   | varchar(11)            | NO   |     | NULL    |       |
+# | datetime     | datetime               | NO   |     | NULL    |       |
+# +--------------+------------------------+------+-----+---------+-------+
+
+READ_PARAMS = textwrap.dedent(
+    """\
+        id, userid, ip, osu_ver, osu_stream, datetime
+    """,
+)
+
+
+class IngameLogin(TypedDict):
+    id: int
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+    datetime: datetime
+
+
+class InGameLoginUpdateFields(TypedDict, total=False):
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+
+
+async def create(
+    user_id: int,
+    ip: str,
+    osu_ver: str,
+    osu_stream: str,
+) -> IngameLogin:
+    """Create a new login entry in the database."""
+    query = f"""\
+        INSERT INTO ingame_logins (userid, ip, osu_ver, osu_stream, datetime)
+             VALUES (:userid, :ip, :osu_ver, :osu_stream, NOW())
+    """
+    params = {
+        "userid": user_id,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    rec_id = await app.state.services.database.execute(query, params)
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = :id
+    """
+    params = {
+        "id": rec_id,
+    }
+    ingame_login = await app.state.services.database.fetch_one(query, params)
+
+    assert ingame_login is not None
+    return cast(IngameLogin, ingame_login)
```

Comment:
```suggestion
    return cast(IngameLogin, dict(ingame_login._mapping))
```

Reply:
pls do for all cases when returning from repos - we adjusted all other cases in #501

Reply:
ingame_login by itself will be a Record instance from aiomysql

ingame_login._mapping gives a Mapping[str, Any], but it's readonly

it's useful to return an updatable model imo, so i typically copy into a new dict